src/palette.c

```diff
@@ -0,0 +1,660 @@
+        }
+    }
+
+    if (startTask == TRUE) {
+        data->selectedBuffer = toSelect;
+        if (!data->callbackFlag) {
+            data->callbackFlag = TRUE;
+            data->selectedFlag = 1;
+            data->forceExit = FALSE;
+            SysTask_CreateOnMainQueue(SysTask_TimedPaletteFade, data, -2);
+        }
+    }
+
+    return startTask;
+}
+
+static u8 IsPaletteSelected(u16 selectedFlag, u16 bufferID) {
+    return (selectedFlag & (1 << bufferID)) != 0;
+}
+
+static void PaletteData_SetTransparentBit(PaletteData *data, u16 bufferID) {
+    if (IsPaletteSelected(data->transparentBit, bufferID) != TRUE) {
+        data->transparentBit |= 1 << bufferID;
+    }
+}
+
+static void MaskOpaqueBit(PaletteBufferId bufferID, PaletteBuffer *buffer, u16 *opaqueBit) {
+    u8 limit;
+    if (bufferID < PLTTBUF_EX_BEGIN) {
+        limit = buffer->size >> 5;
+    } else {
+        limit = buffer->size >> 9;
+    }
+    u16 bits = 0;
+    for (u8 i = 0; i < limit; ++i) {
+        bits += (1 << i);
+    }
+    *opaqueBit &= bits;
+}
+
+static void SelectedPaletteData_SetTimedFadeParams(SelectedPaletteData *selectedBit, u16 opaqueBit, s8 wait, u8 cur, u8 end, u16 nextRGB) {
+    if (wait < 0) {
+        selectedBit->step = abs(wait) + 2;
+        selectedBit->wait = 0;
+    } else {
+        selectedBit->step = 2;
+        selectedBit->wait = wait;
+    }
+    selectedBit->opaqueBit = opaqueBit;
+    selectedBit->cur = cur;
+    selectedBit->end = end;
+    selectedBit->nextRGB = nextRGB;
+    selectedBit->waitStep = selectedBit->wait;
+    if (cur < end) {
+        selectedBit->sign = 0;
+    } else {
+        selectedBit->sign = 1;
+    }
+}
+
+static void SysTask_TimedPaletteFade(SysTask *task, void *taskData) {
+    PaletteData *data = (PaletteData *)taskData;
+
+    if (data->forceExit == 1) {
+        data->forceExit = 0;
+        data->transparentBit = 0;
+        data->selectedBuffer = 0;
+        data->callbackFlag = 0;
+        SysTask_Destroy(task);
+        return;
+    }
+
+    if (data->selectedFlag == 1) {
+        data->transparentBit = data->selectedBuffer;
+        PaletteData_TryApplyScheduledBlendStepToStdBuffersHandleDelay(data);
+        PaletteData_TryApplyScheduledBlendStepToExBuffersHandleDelay(data);
+        if (data->selectedBuffer == 0) {
+            data->callbackFlag = 0;
+            SysTask_Destroy(task);
+            return;
+        }
+    }
+}
+
+void PaletteData_ScheduleFadeTaskEndIfNoSelectedBuffers(PaletteData *data) {
+    if (data->selectedBuffer != 0) {
+        data->forceExit = 1;
+    }
+}
+
+static void PaletteData_TryApplyScheduledBlendStepToStdBuffersHandleDelay(PaletteData *data) {
+    for (u8 i = 0; i < PLTTBUF_EX_BEGIN; ++i) {
+        PaletteData_TryApplyScheduledBlendStepToSingleBufferHandleDelay(data, i, 0x10);
+    }
+}
+
+static void PaletteData_TryApplyScheduledBlendStepToExBuffersHandleDelay(PaletteData *data) {
+    for (u8 i = PLTTBUF_EX_BEGIN; i < PLTTBUF_MAX; ++i) {
```

Comment:
it really doesn't matter much but is there a reason you keep doing ++i instead of i++? It shouldn't matter which one is used in this case

Reply:
It's a practice I carry over from C++, where prefix vs. postfix increment matters more.